Allow merchants to see dispute evidence

brucealdridge opened the pull request:
Fixes #7173 

#### Changes proposed in this Pull Request

Add the ability for merchants to review any evidence that is included with a dispute. 

When a customer disputes an order via their bank or "issuer", evidence can be supplied to the merchant with more information. This can be text-based or a PDF/image. Currently, there is no way for a merchant to view this. This PR is the first step to making the evidence accessible.

Stripe doesn't provide a way for testing `issuer_evidence`, this field is only available on live accounts when provided by the issuer/bank. 

Due to the limitations of testing in development and staging environments. This feature will ship behind a feature flag. This will allow validation of the feature on production accounts.

### Implementation

The issuer_evidence data that is returned via the [API](https://stripe.com/docs/api/disputes/object#dispute_object-issuer_evidence) is structured as below. `text_evidence` is `string|null`, `file_evidence` is an array of up to 5 strings.

```js
{
	evidence_type: 'chargeback',
	file_evidence: [
		'file_xxxxxxxxxxxxx1',
		'file_xxxxxxxxxxxxx2',
	],
	text_evidence: 'Sample text evidence',
}
```

I opted to treat both file-based and text-based evidence as file downloads to standardize the UI and handling of both.
Text-based dispute evidence is able to be clicked and downloaded as a `.txt` file. Any other file-based evidence is downloaded, retaining the filename set on upload.

There are two server REST API's for accessing `file` information from a file_id. One returns file details such as filename, filesize etc. The other returns the file contents as a base64 encoded string. Rather than implement new API's, I opted to utilize the currently available ones for this feature. 

The process of downloading file-based evidence can be slow due to the REST API and the overhead of base64. This means there can be a considerable delay between the user clicking the link and seeing any response. 

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

![image](https://github.com/Automattic/woocommerce-payments/assets/57298/bb99145c-4db0-46d7-9c5a-422e3b8ff153)

#### Testing instructions

As this `issuer_evidence` is not available on test accounts. Any testing of this feature must be done using mocked data. Any string can be supplied for `text_evidence`. To generate valid `file_id`'s I've found the simplest way is to upload merchant evidence against any dispute. These `file_id`'s can be used as `file_evidence` to test the rendering and downloading of these files.

1. Enable the feature flag `_wcpay_feature_dispute_on_transaction_page`
2. On any dispute, upload a file and save (not submit) the evidence.
3. Get the associated file_id. This can be done by reviewing the logs or the dispute object in the developer console.
4. In `client/payment-details/dispute-details/index.tsx` add the following snippet and update `<EvidenceList>` to use this instead of `dispute.issuer_evidence`. 
5. Test varying combinations of `file_evidence` and `text_evidence`

```ts
	const issuerEvidence: IssuerEvidence = {
		evidence_type: 'chargeback',
		file_evidence: [ 'file_xxxxxxxxx' ],
		text_evidence: 'Sample text evidence',
	};
```

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

brucealdridge committed:
Allow merchants to see dispute evidence

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7192` or branch name `add/7173-dispute-evidence` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/7173-dispute-evidence) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a7819b99a2120af91d3d19a91eee0acc0db7859b
- Build time: 2023-09-15 03:31:53 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,85 @@
+	return (
+		<>
+			{ issuerEvidence.file_evidence.map( ( fileId: string, i: any ) => (
+				// eslint-disable-next-line react/jsx-key
+				<FileEvidence fileId={ fileId } />
+			) ) }
```
Nit: rather than eslint-ignore, it's easy enough to add a key here:
```suggestion
				<FileEvidence key={ fileId } fileId={ fileId } />
			) ) }
```

Jinksi commented on client/data/files/types.d.ts:
```diff
@@ -0,0 +1,62 @@
+import ACTION_TYPES from './action-types';
+
+export interface File {
+	/**
```
🏆 These inline type comments are very helpful. Thanks for taking the time to add them!

brucealdridge replied:
Just following your example in previous PR's

brucealdridge committed:
Implement issuer evidence downloads

brucealdridge committed:
Merge branch 'develop' into add/7173-dispute-evidence

brucealdridge commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+			} );
+
+			const link = document.createElement( 'a' );
+			link.href =
+				'data:application/octect-stream;base64,' +
+				downloadRequest.file_content;
```
There doesn't seem to be a straightforward way to create a blob with base64 content. So went this route.

haszari replied:
Clever hack! It works, and doesn't feel _too_ hinky, I think it's shippable for now at least. The important thing is that merchants can see issuer evidence.

brucealdridge commented on client/payment-details/dispute-details/index.tsx:
```diff
@@ -55,6 +56,9 @@ const DisputeDetails: React.FC< DisputeDetailsProps > = ( { dispute } ) => {
 									dispute={ dispute }
 									daysRemaining={ countdownDays }
 								/>
+								<EvidenceList
+									issuerEvidence={ dispute.issuer_evidence }
+								/>
```
This needs to be placed somewhere better, have just put it here for easier testing.

haszari replied:
Seems like the right place to me! It's in the dispute info box, at the bottom. 

- Where is it in the original design? 

I agree though that we should make this a workable implementation (i.e. layout, headings, loading states) in this PR. In your opinion, what do we need to do to get there?

haszari replied:
> Where is it in the original design?

- ✅ It's in `Steps to resolve` step 1, so in the same physical location.
- 🚧 It's in a different component here, since `Steps to resolve` is not ready yet.

##### design
![Screenshot 2023-09-15 at 11 56 31 AM](https://github.com/Automattic/woocommerce-payments/assets/4167300/70c23273-3acf-42b3-a852-c1382555c8b4)

##### this PR
![Screenshot 2023-09-15 at 11 11 55 AM](https://github.com/Automattic/woocommerce-payments/assets/4167300/4c871e58-5de4-401e-8ad2-b033a9770883)

brucealdridge committed:
Fixing commented out type

brucealdridge committed:
cleanup

haszari committed:
Merge branch 'develop' into add/7173-dispute-evidence

haszari commented on client/data/files/types.d.ts:
```diff
@@ -0,0 +1,73 @@
+	 */
+	id: string;
+	/**
+	 * 'dispute_evidence' is expected here, as this is the only type of file upload used by the API at this time.
```
Does this type apply to accessing Stripe-provided (issuer evidence) files too? Or is this just for uploads? If so can name it more specifically: `FileUpload` or `FileUploadMetadata` or `StripeFileUploadMetadata` etc.

brucealdridge replied:
`purpose` is used for all stripe-provided files. This type is currently only used by disputes.

haszari replied:
I'm referring to the type `File` – can it be more specifically named (the comment on `purpose` prompted it!).

brucealdridge replied:
`File` type is similar to `Dispute` or `Charge` type. IMO `FileUpload` would imply that it was only used for uploading.  
Keeping similar language and naming to other APIs made sense to me.

haszari replied:
Understood, I have a pet peeve about `File` because it's such a generic word (unlike `Dispute` and `Charge`). However if these are 1:1 named to match APIs that's a reasonable convention (and something to consider delegating to robots instead of humans).

haszari commented on client/data/files/action-types.ts:
```diff
@@ -0,0 +1,8 @@
+enum TYPES {
+	SET_FILE = 'SET_FILE',
+	SET_ERROR_FOR_FILES = 'SET_ERROR_FOR_FILES',
+}
```
Could these actions have more specific names? What's the scope that they apply to? I see this `files` store is not new, but now that we know more about it maybe we can be more specific with naming. "Files" can mean a lot, but I suspect this has a specific scope & application (accessing/uploading files in Stripe account?)

brucealdridge replied:
I just copied the logic wording from other similar files. TBH I'm a little out of my depth with the action/resolver/selector/hooks files.

haszari replied:
I can relate to that for sure. We should pay attention to that - if the code feels like boilerplate, how can we make it more meaningful to us?

haszari commented on client/data/files/resolvers.ts:
```diff
@@ -0,0 +1,37 @@
+import { ApiError } from 'wcpay/types/errors';
+
+/**
+ * Retrieve a single dispute from the disputes API.
```
Pasted comment alert 🚨

haszari commented on client/data/files/types.d.ts:
```diff
@@ -0,0 +1,73 @@
+	/**
+	 * The file content, base64 encoded.
+	 */
+	file_content: string;
```
A bit weird how file content contains file content – is this named based on a Stripe type? Could call it `FileDownloadInfo` or `FileDownload` maybe.

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+
+interface Props {
+	issuerEvidence: IssuerEvidence | null;
+}
```
Should this type be declared closer to where it's used, or have a more descriptive name? Is it a one-off – i.e. only used when declaring that component (`EvidenceList`)?

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+const TextEvidence: React.FC< {
+	evidence: string;
+} > = ( { evidence } ): JSX.Element => (
+	// eslint-disable-next-line jsx-a11y/anchor-is-valid
```
Reminder – add a comment explaining why it's ok to ignore.

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+		className="dispute-evidence-link"
+		download="evidence.txt"
+	>
+		{ __( 'Evidence.txt', 'woocommerce-payments' ) }
```
Might be useful to have a translation hint here – this is a default file name.

Side note, is it practical to add disambiguating info to the filename, e.g. order number, payment id, or dispute id?

brucealdridge replied:
> Might be useful to have a translation hint here – this is a default file name.

Updated.

> Side note, is it practical to add disambiguating info to the filename, e.g. order number, payment id, or dispute id?

I like this idea, will just need to add an extra argument to `IssuerEvidenceList` and `TextEvidence`

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+	const { file, isLoading } = useFiles( fileId );
+	const { createNotice } = useDispatch( 'core/notices' );
+
+	const onDownload = async ( e: React.MouseEvent< HTMLAnchorElement > ) => {
```
I'd name this `onClick`, since it's an onClick handler (confused me for a minute).

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+						{ file?.title || file.filename }
+					</a>
+				) : (
+					<></>
```
Curious if this fallback should be inside or outside `Loadable` – if there's no file id, there's no load wait? (not a blocker, just curious)

brucealdridge replied:
This is for the case when the API doesn't return `200` response. I have not tested.

haszari replied:
Right, so it's an exceptional fallback, rather than an expected state. Maybe could null out.

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+					<a
+						href="#"
+						className="dispute-evidence-link"
+						onClick={ onDownload }
```
If we're fully overriding normal link behaviour (`preventDefault`, render a hidden download "button" and click it), I don't think we should render as a link. 

If we used a button or other "action" ui then the behaviour would feel more consistent. Could even use a button with a built in "preparing your download" state.

brucealdridge replied:
I've updated this to use a button rather than a link.

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+	);
+};
+
+const EvidenceList: React.FC< Props > = ( { issuerEvidence } ) => {
```
Is this just `IssuerEvidenceList` or do you anticipate use for other (merchant) evidence in future? Suggest more specific name.

I'd name based on function … e.g. `IssuerEvidenceDownload` or `IssuerEvidenceFiles`. 

(minor/nitpick)

brucealdridge replied:
I've renamed it to `IssuerEvidenceList`. This component wouldn't be reused.

The `FileEvidence` component could be reused on the merchant evidence form.

haszari commented on client/payment-details/dispute-details/evidence-list.tsx:
```diff
@@ -0,0 +1,117 @@
+	) {
+		return (
+			<span>
+				{ __( 'No evidence available', 'woocommerce-payments' ) }
```
Is this going to happen in practice? Might need to review this string. Suggest adding `issuer` there for clarity, so could be `No issuer evidence available` or `No evidence provided by issuer`.

Include a screenshot of this state for design review.

brucealdridge replied:
Not sure. This is one of the many unknowns that we need to test on production. I've updated to render nothing when there is no evidence for now.

haszari reviewed (approved):
This is looking good and I want to ship it!

I left a few comments on the code, no major issues.

The one blocker from my point of view is tidying up the UI a little to make it consistent and shippable. Here's one rough hack design idea.

- Use buttons since the behaviour is an action: "request download", not a link. I'm hoping we have an action button with a "one moment please" spinner/progress indication which would be a good match for behaviour.
- Add a heading so it's clear what these things are – issuer evidence.
- Hide buttons in a disclosure triangle so we don't see them slowly load individually.

Keen to hear input / ideas from @nikkivias of course :)

brucealdridge committed:
UI improvments

brucealdridge committed:
Add docblocks to new API methods

brucealdridge committed:
Rename component to IssuerEvidenceList

brucealdridge committed:
Replace links with buttons

brucealdridge committed:
Changes from PR review

brucealdridge committed:
Adding changelog

Commit:
style tidies for dispute issuer evidence:
- borrow wc list header style for title
  - bonus: sentence case for title
- appropriate padding between title & evidence buttons

Commit:
Merge branch 'add/7173-dispute-evidence' of github.com:Automattic/woocommerce-payments into add/7173-dispute-evidence

brucealdridge committed:
Update title for issuer evidence

brucealdridge committed:
Update issuer evidence title

Commit:
dispute issuer evidence – use WP PanelBody so buttons can be hidden (rough cut)

Commit:
Merge branch 'add/7173-dispute-evidence' of github.com:Automattic/woocommerce-payments into add/7173-dispute-evidence

brucealdridge committed:
removing unneeded type

Commit:
reinstate panelbody styling (accidentally ate in merge)

Commit:
more consistent, spacious padding for evidence buttons

Commit:
Merge branch 'add/7173-dispute-evidence' of github.com:Automattic/woocommerce-payments into add/7173-dispute-evidence

haszari committed:
Merge branch 'develop' into add/7173-dispute-evidence

Commit:
remove extraneous padding so dispute issuer evidence is snug in container

brucealdridge commented:
Manually merged due to GitHub actions being stuck.
